R4R: lamborghini distribution & inflation spec upgrade

rigelrozanski opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes. 
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  --> 
closes #967 

* [x] Updated all relevant documentation (`docs/`)
* [x] Updated all relevant code comments
* [x] Wrote tests
* [x] Updated `CHANGELOG.md`
* [x] Updated `cmd/gaia` and `examples/`
___________________________________
For Admin Use:
* [x] Added appropriate labels to PR (ex. wip, ready-for-review, docs)
* [ ] Reviewers Assigned 
* [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

rigelrozanski commented on docs/spec/distribution/end_block.md:
```diff
@@ -0,0 +1,31 @@
+and global pool.  When the validator is the proposer of the round, that
+validator (and their delegators) receives between 1% and 5% of fee rewards, the
+reserve tax is then charged, then the remainder is distributed socially by
+voting power to all validators including the proposer validator.  The amount of
```
All bonded validators. Not just ones that voted, the only way to not get the social distribution is to get kicked out (which will happen if you're not live enough) - I'll make this text more explicit.

rigelrozanski replied:
Added more text to make this explicit

alexanderbez commented on docs/spec/inflation/end_block.md:
````diff
@@ -0,0 +1,57 @@
+# End Block
+
+```
+EndBlock() 
````
What are these for?

rigelrozanski replied:
Unnecessary - removing

alexanderbez commented on docs/spec/inflation/end_block.md:
```diff
@@ -0,0 +1,57 @@
+
+Validator provisions are minted on an hourly basis (the first block of a new
+hour). The annual target of between 7% and 20%. The long-term target ratio of
+bonded tokens to unbonded tokens is 67%.
```
Would it be possible to describe why the exact target of 67% in a concise manner?

cwgoes replied:
Agreed - or cite the whitepaper / relevant source.

rigelrozanski replied:
I disagree this IS the spec, it supersedes the white paper - it's not a place where choices always need to be justified. This number for instance comes out of a bunch of conversations with Jae - there is no good reason it just seems like a fine value. For y'alls record if the value is to small its a security concern, if it's to large then there isn't enough liquidity for price discovery -> 67% seems fine.

alexanderbez commented on docs/spec/inflation/end_block.md:
```diff
@@ -0,0 +1,57 @@
+        inflation = nextInflation(hrsPerYr).Round(precision)
+        SetInflation(inflation)
+        
+        provisions = inflation * (pool.TotalSupply() / hrsPerYr)
```
Is the `pool.TotalSupply()` here the total number of bonded tokens in the validator set?

cwgoes replied:
No, it's the total token supply - all tokens bonded, unbonded, deposited-in-governance-proposals, etc.

rigelrozanski replied:
comment addressed by cwgoes

alexanderbez commented on docs/spec/inflation/end_block.md:
```diff
@@ -0,0 +1,57 @@
+
+    bondedRatio = pool.BondedPool / pool.TotalSupply()
+
+    inflationRateChangePerYear = (1 - bondedRatio / params.GoalBonded) * params.InflationRateChange
```
What exactly is `params.InflationRateChange`?

cwgoes replied:
Rate of change in the inflation rate (second derivative of the total supply)

rigelrozanski replied:
comment addressed by cwgoes

cwgoes commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+ - Multi-token fees to be socially distributed 
+ - Proposer reward pool
+ - Inflated atom provisions
+ - Validator commission on delegators 
```
Commission on atom provisions from delegated stake?

rigelrozanski replied:
Updated to `commission on all rewards earned by their delegators`

cwgoes commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+
+## Overview
+
+Collected fees are pooled globally and divided out passively to validators and
```
Why passively? (e.g. "Fee distribution is lazy so that we don't need to modify every account which will receive fees every block")

rigelrozanski replied:
yes passively meaning that the computations take place at withdrawal (and less calculations required)

cwgoes replied:
Right, I think we should include this rationale, since it might not be initially obvious to someone who hasn't thought through fee distribution to the extent we have.

cwgoes commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+## Affect on Staking
+
+Commission on Atom provisions and having autobonded-atom-provisions have been
+shown to be  mutually exclusive. Fundamentally if there are atoms commissions
```
Can we cite the longer-form previous discussion here?

rigelrozanski replied:
I think this is already sufficient as it contains the condensed points technical points as to why these two mechanisms are mutually exclusive. note also that this understanding did not come out of a larger discussion with the community in forums or on github but from in person conversations in Berkley.
(ps. I've updated this paragraph to be a bit clearer)

cwgoes replied:
New paragraph is clearer, thanks.

cwgoes commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+block making a withdrawal of fees for a delegator require a calculation for
+every single block since the last withdrawal. In conclusion we can only have atom
+commission and unbonded atoms provisions, or bonded atom provisions and no atom
+commission. 
```
", and we elect to implement the former. Stakeholders wishing to rebond their provisions may elect to set up a script to periodically withdraw and rebond fees."

rigelrozanski replied:
added

cwgoes commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+Note that the reward pool is held decimal coins (`DecCoins`) to allow 
+for fractions of coins to be received from operations like inflation. 
+When coins are distributed from the pool they are truncated back to 
+`DecCoins` which is non-decimal. 
```
I also find this naming confusing. Maybe "NatCoins" or "WholeCoins" (or just "Coins")?

rigelrozanski replied:
This was a typo - the final `DecCoins` should actually just be `sdk.Coins`.

cwgoes commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+When coins are distributed from the pool they are truncated back to 
+`DecCoins` which is non-decimal. 
+
+ - Global:  `0x00 -> amino(global)`
```
A store prefix I think.

rigelrozanski replied:
This is the standard that the spec is using which bucky defined or is following? (this is why the spec for a spec that is needed) - in this situation the byte actually represents the key in the store NOT the prefix

cwgoes replied:
The key in the store? I don't follow. The end `byte[]` store key is something like

```golang
append(stakeKey, append(byte(0x00)))
```

right?

cwgoes commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+type Global struct {
+	PrevBondedTokens  sdk.Dec  // bonded token amount for the global pool on the previous block 
+    Adjustment        sdk.Dec  // global adjustment factor for lazy calculations
+    Pool              DecCoins // funds pool for all validators
```
Funds yet to be withdrawn?

rigelrozanski replied:
thank you added

cwgoes commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+    Pool              DecCoins // funds pool for all validators
+    PrevReceivedPool  DecCoins // funds added to the pool on the previous block
+    EverReceivedPool  DecCoins // total funds ever added to the pool 
+    CommunityFund     DecCoins // pool for community funds
```
Yet to be spent? How can this be spent from (governance?)

rigelrozanski replied:
added - and correct / correct

cwgoes commented on docs/spec/distribution/state.md:
````diff
@@ -0,0 +1,108 @@
+
+```golang
+type ValidatorDistribution struct {
+    CommissionWithdrawalHeight   int64    // last time this delegation withdrew rewards
````
This delegation or this validator?

rigelrozanski replied:
changed to `last time this validator withdrew commission`

cwgoes commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+    WithdrawalHeight   int64    // last time this delegation withdrew rewards
+    Adjustment         sdk.Dec  // commission adjustment factor
+    AdjustmentProposer DecCoins // reward pool collected from being the proposer
+    PrevTokens         sdk.Dec  // bonded tokens held by the delegation on the previous block
```
Are these collected fees? Or do you mean "of the delegation itself" - in that case, why store these in addition to the delegation object?

rigelrozanski replied:
Assuming you're talking about the comments on the adjustment factors, the comments were totally incorrect referring to something else (copy paste error). updated to:
``` golang
    Adjustment         sdk.Dec  // fee provisioning adjustment factor
    AdjustmentProposer DecCoins // proposers pool adjustment factor
```

cwgoes commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+other validators to run through. Each power change is stored under a sequence
+number which increments by one for each new power change record
+
+ - PowerChange: `0x03 | amino(PCSequence) -> amino(validatorDist)`
```
Should we prefix this key with block height? Will make epoch-based pruning easier later.

rigelrozanski replied:
huh - I never thought of that - I think we can easily JUST prefix it with the block height don't even need the PCSequence - cool! updating - also removing max sequence as nolonger needed woot 👏

cwgoes commented on docs/spec/distribution/transactions.md:
```diff
@@ -0,0 +1,396 @@
+## TxWithdrawDelegation
+
+When a delegator wishes to withdraw their transaction fees it must send
+`TxWithdrawDelegation`. Note that parts of this transaction logic is also
```
"are also"

rigelrozanski replied:
addressed

cwgoes commented on docs/spec/distribution/transactions.md:
```diff
@@ -0,0 +1,396 @@
+triggered each with any change in individual delegations, such as an unbond,
+redelegation, or delegation of additional tokens to a specific validator.  
+
+Generally, each time a withdrawal is made by a recipient the adjustment term
```
"Generally" - are there exceptions?

rigelrozanski replied:
There are not, removed `Generally`

cwgoes commented on docs/spec/distribution/transactions.md:
````diff
@@ -0,0 +1,396 @@
+
+```golang
+type TxWithdrawDelegation struct {
+    delegatorAddr sdk.AccAddress
````
Do we want the ability to specify *which* delegation is delegator is withdrawing from?

They could have multiple delegations to separate validators (or even to the same validator), and might want to withdraw at separate times (or to avoid using too much gas).

rigelrozanski replied:
probably not something we want anytime in the near future - I _think_ that there are some shared computations when a delegator withdraws all of their tokens as opposed to doing this through multiple transactions.  - Due to how things are structured internally we would be able to introduce this down the road, but I think for now let's stick with the obviously required transactions of processing them all simultaneously.

cwgoes replied:
Makes sense :+1:

cwgoes commented on docs/spec/distribution/transactions.md:
```diff
@@ -0,0 +1,396 @@
+
+## TxWithdrawValidator
+
+When a validator wishes to withdraw their transaction fees it must send
```
To clarify - this is *not* the same as withdrawing fees from the self-bond, which must be done separately, correct?

cwgoes commented on docs/spec/distribution/transactions.md:
```diff
@@ -0,0 +1,396 @@
+for that distributor had static shares (equal to the current shares), and no
+recipients had ever withdrawn their entitled rewards. The projected pool
+represents the anticipated recipient's entitlement to the distributors tokens
+base on the current blocks token input to the distributor, and the
```
"token input" = "contributing stake"?

cwgoes commented on docs/spec/distribution/transactions.md:
```diff
@@ -0,0 +1,396 @@
+# Transactions
```
I find it really difficult to reason through the correctness of the transaction procedures in this file, likely because we're trying to implement a simple strict model with a lazy process.

I think it would be helpful if we outline for each transaction what the calculation *would be* in the strict, inefficient case (iterating over everything), *then* explain our lazy implementation, and even ideally provide a quick (rough) demonstration of equivalence.

rigelrozanski replied:
Coming up with a good way to comprehend this efficient fee distribution module is going to be an ongoing challenge - Due to this, I agree that equivalent outcome in a "simple" module for each transaction should be documented within the spec beside the complex "efficient" situation already provided.

With regards to the example - this is a tug of worlds here - I don't think that we want to have non-spec information in the spec - to me, the example described doesn't actually belong in the spec itself HOWEVER I think it's super valuable and I think we should have it somewhere.  I think maybe IN The spec we could have a non-spec section which actually works through what you're describing.  - I think that this is a separate task which should be completed soon, but is a lower priority than distribution implementation to me

cwgoes replied:
I don't mean include an example (although that might be a good idea too), I mean include equations that demonstrate why the complex implementation reduces to the same effects as the simple one (weren't those written up somewhere previously)?

cwgoes replied:
Or "parameterized examples"!

rigelrozanski replied:
Leaving this for now as we're likely moving away from pre-launch implementation however here is an issue for more discussion https://github.com/cosmos/cosmos-sdk/issues/1906

cwgoes commented on docs/spec/distribution/triggers.md:
```diff
@@ -0,0 +1,31 @@
+
+## Create validator distribution
+
+ - triggered-by: `stake.TxCreateValidator`
```
Every time or only when the new validator candidate is bonded?

cwgoes replied:
(and in the latter case - then also by another validator unbonding, for example?)

cwgoes replied:
Also, do we need to anything when a validator is *unbonded*?

rigelrozanski replied:
>Every time or only when the new validator candidate is bonded?

Good observation, this only happens when the validator enters the bonded validator set. updated.

> Also, do we need to anything when a validator is unbonded?

**For other validators:**
When a validator is unbonded - the power change is be recorded and thus all other's passive calculations will incorporate the changes passively as expected. So nothing additional or special is required for an unbond for them. 

**For the affected validator:**
With regards to everyone affected by the unbond - we will need to incorporate a new mechanism to passively account for this - Just opened up an issue to think about this a bit more https://github.com/cosmos/cosmos-sdk/issues/1905.
I'm certain that the appropriate mechanism can be added onto what is spec'd here to account for this situation. THANK YOU FOR BRINGING THIS UP.

cwgoes commented on docs/spec/distribution/triggers.md:
```diff
@@ -0,0 +1,31 @@
+# Triggers
```
What about slashing events (which modify validator stake and delegation stake, and not necessarily in the same proportion)?

cwgoes replied:
Also, what about parameter changes (e.g. enacted by `ParameterChangeProposal`s)?

rigelrozanski replied:
Slashing events is a situation that needs to be considered:

Side note: Equal slashing of delegators and validators is a design constraint of this system - we can talk about it more outside of github if you'd like. 

These slashing events will need to have a new special computation not yet outlined in this spec: can be processed similarly with a new pool as per to unbonding validators. more discussion here: https://github.com/cosmos/cosmos-sdk/issues/1905

________
Parameter changes will not trigger any new changes because parameters relevant to distribution affect the tokens flowing into the distribution system but do not affect the tokens already contained within this distribution system.

cwgoes commented on docs/spec/distribution/triggers.md:
```diff
@@ -0,0 +1,31 @@
+
+## Create or modify delegation distribution
+ 
+ - triggered-by: `stake.TxDelegate`
```
Or `TxBeginRedelegate`, or `TxBeginUnbonding`?

rigelrozanski replied:
included

ValarDragon commented:
I'd like to review this as well before its merged. Chris had recommended that I wait until his comments were resolved first though. Commenting so I get pinged for updates.

rigelrozanski commented:
Hopefully can address everything tomorrow -still focused on decimal

rigelrozanski commented on PENDING.md:
```diff
@@ -38,6 +38,7 @@ IMPROVEMENTS
 * [tools] Remove `rm -rf vendor/` from `make get_vendor_deps`
 * [x/auth] Recover ErrorOutOfGas panic in order to set sdk.Result attributes correctly
 * [x/stake] Add revoked to human-readable validator 
+* [spec] Inflation and distribution specs drastically improved
```
addressed

rigelrozanski commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+delegators. Each validator has the opportunity to charge commission to the
+delegators on the fees collected on behalf of the delegators by the validators.
+Fees are paid directly into a global fee pool, and validator proposer-reward
+pool. Due to the nature of of passive accounting whenever changes to parameters
```
addressed

rigelrozanski commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+which affect the rate of fee distribution occurs, withdrawal of fees must also
+occur. 
+ 
+ - when withdrawing one must withdrawal the maximum amount they are entitled
```
addressed (lower bullet points from here - using lower case in bullet points be cause not sentences.

rigelrozanski commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+ - when a validator chooses to change the commission on fees, all accumulated 
+   commission fees must be simultaneously withdrawn.
+
+The above scenarios are covered is `triggers.md`.
```
lol thanks - was a type `is` -> `in`

rigelrozanski commented on docs/spec/distribution/future_improvements.md:
```diff
@@ -0,0 +1,16 @@
+### Power Change
+
+Within the current implementation all power changes ever made are indefinitely stored
+within the current state. In the future this state should be trimmed on an epoch basis. Delegators
```
We need all the calculations in the current state for an epoch equal to what we want the maximum allowable time withdrawal your fees is.. I'm guessing at minimum this is something like a couple months and at maximum is something like a year. - There is some complexity surrounding tracking what the last required power change is but the bigger problem is figuring out what to do / the appropriate lazy mechanism which plays nice with the existing fees distribution mechanism for everyone who hasn't withdrawn their fees by that time. There are a number of options: But this is the discussion grounds for that:

cwgoes replied:
Agreed that this merits discussion but it probably can be safely tabled to post-launch as long as we put in place the store key prefixes (block height) to prune easily in the future.

rigelrozanski commented on docs/spec/distribution/end_block.md:
```diff
@@ -0,0 +1,31 @@
+validator (and their delegators) receives between 1% and 5% of fee rewards, the
+reserve tax is then charged, then the remainder is distributed socially by
+voting power to all validators including the proposer validator.  The amount of
+proposer reward is calculated from pre-commits Tendermint messages. All
```
sure, included.

rigelrozanski commented on docs/spec/distribution/end_block.md:
```diff
@@ -0,0 +1,31 @@
+                       * sumPowerPrecommitValidators / totalBondedTokens)
+     proposer.ProposerPool += proposerReward
+     
+     communityFunding = feesCollectedDec * communityTax
```
leaving for now - open to suggestions still

rigelrozanski commented on docs/spec/distribution/end_block.md:
```diff
@@ -0,0 +1,31 @@
+At each endblock, the fees received are sorted to the proposer, community fund,
+and global pool.  When the validator is the proposer of the round, that
+validator (and their delegators) receives between 1% and 5% of fee rewards, the
+reserve tax is then charged, then the remainder is distributed socially by
```
Well I did actually mean social distribution which means that you get coins even if you weren't the proposer - but proportional makes more sense in this instance, so I'll change, but then name is the social distribution

cwgoes replied:
OK makes sense - if you just define "social distribution" once that's fine.

rigelrozanski commented on docs/spec/distribution/overview.md:
```diff
@@ -0,0 +1,50 @@
+
+## Affect on Staking
+
+Commission on Atom provisions and having autobonded-atom-provisions have been
```
Yeah it's a mouthful - Although it was accurate, I've expanded some of the sentences, making them a bit more verbose but also hopefully a bit more palatable

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+
+### Global
+
+All globally tracked parameters for distribution are stored within
```
Thanks, addressed. Just a note about github and PR comments - I couldn't actually see the line you were referring to because the comment was left on the line above it.

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+`Global`.  Rewards are collected are added to the reward pool and
+distributed to validators/delegators from here. 
+
+Note that the reward pool is held decimal coins (`DecCoins`) to allow 
```
thanks, addressed

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+}
+
+type Global struct {
+	PrevBondedTokens  sdk.Dec  // bonded token amount for the global pool on the previous block 
```
addressed was a tab instead of spaces so i didn't see in my editor (which is 4spaces to a tab)

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+
+Validator Distribution information for the relevant validator is updated: each
+time delegations to a validator are updated, a validator successfully proposes
+a block and receives (receiving a reward), any delegator withdraws from a
```
thanks addressed

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+
+Validator Distribution information for the relevant validator is updated: each
+time delegations to a validator are updated, a validator successfully proposes
+a block and receives (receiving a reward), any delegator withdraws from a
```
dup comment (already addressed)

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+
+### Validator Distribution
+
+Validator Distribution information for the relevant validator is updated: each
```
lol I actually just put these items in a list write before reading this comment - totes. Also change to `updated each time`:

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+    ProposerPool               DecCoins  // reward pool collected from being the proposer
+    EverReceivedProposerReward DecCoins  // all rewards ever collected from being the proposer
+    PrevReceivedProposerReward DecCoins  // all rewards ever collected from being the proposer
+	PrevBondedTokens           sdk.Dec   // bonded token amount on the previous block 
```
darn tabs (thx) - went through and updated all the padding

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+    ProposerAdjustment         DecCoins  // proposer pool adjustment factor
+    ProposerPool               DecCoins  // reward pool collected from being the proposer
+    EverReceivedProposerReward DecCoins  // all rewards ever collected from being the proposer
+    PrevReceivedProposerReward DecCoins  // all rewards ever collected from being the proposer
```
correct thanks

rigelrozanski commented on docs/spec/distribution/state.md:
```diff
@@ -0,0 +1,108 @@
+### Delegation Distribution 
+
+Each delegation holds multiple adjustment factors to specify its entitlement to
+the rewards from a validator. `AdjustmentPool` is  used to passively calculate
```
yes - the first sentence is the summary, the next two describe the fields and specifies mechanism more.

rigelrozanski commented on docs/spec/distribution/triggers.md:
```diff
@@ -0,0 +1,31 @@
+
+ - triggered-by: `stake.TxCreateValidator`
+
+Whenever a totally new validator is added to the Tendermint validator set they
```
lol yeah totally - just my valley bleeding in.

rigelrozanski commented on docs/spec/distribution/triggers.md:
```diff
@@ -0,0 +1,31 @@
+ - triggered-by: `stake.TxCreateValidator`
+
+Whenever a totally new validator is added to the Tendermint validator set they
+are entitled to begin earning rewards of atom provisions and fess. At this
```
addressed

cwgoes commented:
Per discussion with @rigelrozanski we'd like to consider the "piggy bank distribution" instead as a simpler approach for launch.

rigelrozanski commented:
Yes more on that - this spec should still be merged and reviewed further and expanded based on some of the new issues I've raised out of this. This model is still something we want - but should likely NOT be a prelaunch requirement - instead we should use the simple piggy bank model which should be spec'd out separately.  

@cwgoes  - I've addressed your comments, although a few of them haven't been fully addressed and I've opened two github issues for further consideration on this mechanism. 
@alexanderbez  - I've addressed all of your comments besides adding examples - again based on this comment - that's something we should do PRE implementation - however I'm going to hold off on that for now because I don't think we should implement this prelaunch 
@ValarDragon - Probably not a good use of your time to focus on reviewing this at the moment because I think we should move away from prelaunch implementation of this spec

rigelrozanski commented:
Let's merge this?

cwgoes commented:
> Let's merge this?

Before we actually implement it in the future, I want to review again re: transaction correctness - but agreed, fine for now.
